src/osd/PrimaryLogPG.cc

```diff
@@ -7727,14 +7728,20 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
 	bufferlist bl;
 	if (oi.is_omap()) {
 	  ObjectMap::ObjectMapIterator iter = osd->store->get_omap_iterator(
-	    ch, ghobject_t(soid)
+	    ch, ghobject_t(soid),
+	    // try to seek as many keys-at-once as possible for the sake of performance.
+	    // note complexity should be logarithmic, so seek(n/2) + seek(n/2) is worse
+	    // than just seek(n).
+	    filter_prefix >= start_after ? filter_prefix : start_after
 	    );
           if (!iter) {
             result = -ENOENT;
             goto fail;
           }
-	  iter->upper_bound(start_after);
-	  if (filter_prefix > start_after) iter->lower_bound(filter_prefix);
+	  if (filter_prefix < start_after) {
+	    // the get_omap_iterator() seeks to lower bound
+	    iter->upper_bound(start_after);
```

Comment:
In this case we have already executed a `lower_bound` in OmapIterator constructor.

Only initial omap listings that provide `filter_prefix` will be exempt from calling it. It does not feel like a corner case.

Reply:
The `upper_bound()` here moves the iterator's position up to 1 key. I assume almost entire work is performed in `get_omap_iterator()`; here we basically do `next()`.

The profiled case from the field was listing with non-empty `filter_prefix` and `max_keys=2`.

Reply:
Anyway, get rid of this 1-key-bumps entirely in the revision pushed yesterday.